staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -741,7 +742,15 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 				return storage.NewInternalErrorf("unable to transform key %q: %v", kv.Key, err)
 			}
 
-			if err := appendListItem(v, data, uint64(kv.ModRevision), opts.Predicate, s.codec, s.versioner, newItemFunc); err != nil {
+			// Check if the request has already timed out before decode object
+			select {
+			case <-ctx.Done():
+				// parent context is canceled or timed out, no point in continuing
+				return storage.NewTimeoutError(string(kv.Key), "request did not complete within requested timeout")
+			default:
+			}
+
+			if err := appendListItem(ctx, v, data, uint64(kv.ModRevision), opts.Predicate, s.codec, s.versioner, newItemFunc); err != nil {
```

Comment:
Sorry, I don't quite understand the meaning of `not record every object separately`, may you elaborate on that?

Reply:
Currently, you're recording decoding of each item separately.
I was wondering if we don't want to record it once per the full page.

But I guess the answer is no, as it would also contain transformFromStorage, which can be fairly long operation...

Reply:
Yeah, there's no way we can bypass Transform to make it work.